File: Python/Algorithms/Sorting-Algorithms/merge_sort.py
# time complexity = O(nlogn)
# space complexity = O(n) since we need to store the merged array b in merge()
# merge sort uses divide and conquer
# we keep on dividing an array along the middle index into adjacent sub-arrays and sorting them recursively

# function to merge 2 sorted sub-arrays/halves of array a
# sort_order is used to determine if the array will be sorted in ascending or descending order
def merge(a, start, mid, end, sort_order):
	i = start # starting index of subarray 1
	j = mid + 1 # starting index of subarray 2
	b = []
  # compare i-th and j-th element sub-array 1 and 2 respectively
  # append whichever is smaller or both if both are equal
  # loop break when we reach the end or one or both of the sub arrays
	while i <= mid and j <= end:
		if sort_order*a[i] < sort_order*a[j]:
			b.append(a[i])
			i += 1
		elif sort_order*a[j] < sort_order*a[i]:
			b.append(a[j])
			j += 1
		else:
			b.append(a[i])
			i += 1
			b.append(a[j])
			j += 1
	# if any element is left in sub array1, append to b
	while i <= mid:
		b.append(a[i])
		i += 1
	# if any element is left in sub array2, append to b
	while j <= end:
		b.append(a[j])
		j += 1
	
	# copy the sorted array into a
	# subarray 1 and subarray 2 are adjacent hence we just sorted array a from index start to end
	for i in range(start, end+1):
		a[i] = b[i - start]


def mergeSort(a, start, end, sort_order=1):
	# if start == end i.e. sort one element
	# array of one element is always sorted
	if start != end:
		mid = (start + end) // 2
		# sort subarray 1 - left half
		mergeSort(a, start, mid, sort_order)
		# sort subarray 2 - right half
		mergeSort(a, mid+1, end, sort_order)
		# merge sorted subarrays/ halves
		merge(a, start, mid, end, sort_order)

# driver code
if __name__ == "__main__":
	sort_order = input("Enter 'd' for descending order of sort (by default ascending order is selected): ")
	l = list(map(int, input("Enter list to sort: ").split(" ")))

	print("Before sorting:", l)
	if sort_order == 'd':
		mergeSort(l, 0, len(l) - 1, -1)
		print("Sorting in descending order...")
	else:
		mergeSort(l, 0, len(l) - 1)
		print("Sorting in ascending order...")
	print("After sorting:", l)

File: C++/Algorithms/Sorting-Algorithms/MergeSort/Merge_Sort.cpp
/*
Merge Sort 
The list is divided into lists until the length is 0 or 1, then the partial lists are recursively sorted using merge sort, merged again, and repeated until the initial list length is reached.
Time Complexity(worst/averaage/Best) : (n*log(n))
Space Complexity: n
Contributor : Ankit Surwade
*/

#include <bits/stdc++.h>
using namespace std;

void merge(int arr[], int low, int mid, int high){
  int i, j, k;
  int n1 = mid + 1 - low; 
  int n2 = high - mid;
  int l[n1], r[n2];
  for(int i=0; i<n1; i++) l[i]=arr[low+i]; // left
  for(int j=0; j<n2; j++) r[j]=arr[mid+j+1]; // right
  i=0; j=0; k=low;

  while(i<n1 && j<n2){
    if(l[i]<=r[j]) arr[k++]=l[i++];
    else arr[k++]=r[j++];
  }
  while(i<n1) arr[k++]=l[i++];
  while(j<n2) arr[k++]=r[j++];
}

void mergeSort(int arr[],int low,int high){ // sorting 
  int mid;
  if(low<high){
    mid=low+(high-low)/2;

    mergeSort(arr,low,mid);
    mergeSort(arr,mid+1,high);
    merge(arr,low,mid,high);
  }
}  

int main()
  {
    //enter number of elements
    int n;
    cin>>n;
    int arr[n];
    //enter elemetns
    for(int i=0;i<n;i++) cin>>arr[i];
    //called Function
    mergeSort(arr,0,n-1);
    //after Sort array
    for(int i=0;i<n;i++) cout<<arr[i]<<" ";

    return 0;
  }
